Substitutions as part of yaml key for actions stopped working for dev

### The problem

I'm not sure if it's a bug or a feature, but for some reason in a project I'm using actions for example in this way:
```
substitutions:
  default_solar_sell: "on" # "on", "off"
```
```
script:
  - id: set_safe_modbus_registers
    then:
      - logger.log: 
          level: INFO
          format: Setting safe Modbus register values
      - switch.turn_${default_solar_sell}:
          id: switch_solar_sell
```

And I have noticed that it stopped working for dev. It's still working correctly for stable and beta.

In my pipeline during build I'm getting the following error:
```
Failed config

script: [source pv_inverter/deye_hybrid_3p_lv.yaml:40]
  - id: set_safe_modbus_registers
    then: 
      - logger.log: 
          level: INFO
          format: Setting safe Modbus register values
      
      Key missing from action! Got {}.
      - {}
Compile firmware
Error: Process completed with exit code 2.
```


### Which version of ESPHome has the issue?

2025.7.0-dev and 2025.7.0b1

### What type of installation are you using?

Home Assistant Add-on

### Which version of Home Assistant has the issue?

_No response_

### What platform are you using?

ESP32

### Board

_No response_

### Component causing the issue

_No response_

### YAML Config

```yaml
substitutions:
  default_solar_sell: "on" # "on", "off"

script:
  - id: set_safe_modbus_registers
    then:
      - logger.log: 
          level: INFO
          format: Setting safe Modbus register values
      - switch.turn_${default_solar_sell}:
          id: switch_solar_sell
```

### Anything in the logs that might be useful for us?

```txt
Failed config

script: [source pv_inverter/deye_hybrid_3p_lv.yaml:40]
  - id: set_safe_modbus_registers
    then: 
      - logger.log: 
          level: INFO
          format: Setting safe Modbus register values
      
      Key missing from action! Got {}.
      - {}
Compile firmware
Error: Process completed with exit code 2.
```

### Additional information

_No response_

--- a/esphome/components/substitutions/__init__.py
+++ b/esphome/components/substitutions/__init__.py
@@ -151,8 +151,11 @@ def _substitute_item(substitutions, item, path, jinja, ignore_missing):
             if sub is not None:
                 item[k] = sub
         for old, new in replace_keys:
-            item[new] = merge_config(item.get(old), item.get(new))
-            del item[old]
+            if str(new) == str(old):
+                item[new] = item[old]
+            else:
+                item[new] = merge_config(item.get(old), item.get(new))
+                del item[old]
     elif isinstance(item, str):
         sub = _expand_substitutions(substitutions, item, path, jinja, ignore_missing)
         if isinstance(sub, JinjaStr) or sub != item:

--- a/tests/unit_tests/fixtures/substitutions/00-simple_var.approved.yaml
+++ b/tests/unit_tests/fixtures/substitutions/00-simple_var.approved.yaml
@@ -17,3 +17,5 @@ test_list:
   - ${undefined_var}
   - $undefined_var
   - ${ undefined_var }
+  - key1: 1
+    key2: 2

--- a/tests/unit_tests/fixtures/substitutions/00-simple_var.input.yaml
+++ b/tests/unit_tests/fixtures/substitutions/00-simple_var.input.yaml
@@ -19,3 +19,5 @@ test_list:
   - ${undefined_var}
   - $undefined_var
   - ${ undefined_var }
+  - key${var1}: 1
+    key${var2}: 2

--- a/tests/unit_tests/fixtures/substitutions/03-closures.approved.yaml
+++ b/tests/unit_tests/fixtures/substitutions/03-closures.approved.yaml
@@ -6,6 +6,7 @@ package_result:
     root file
   - Double substitution also works; the value of var7 is 79, where A is a package
     var
+  - key79: Key should substitute to key79
 local_results:
   - The value of B is 5
   - 'You will see, however, that

--- a/tests/unit_tests/fixtures/substitutions/closures_package.yaml
+++ b/tests/unit_tests/fixtures/substitutions/closures_package.yaml
@@ -1,3 +1,4 @@
 package_result:
   - The value of A*B is ${A * B}, where A is a package var and B is a substitution in the root file
   - Double substitution also works; the value of var7 is ${var$A}, where A is a package var
+  - key${var7}: Key should substitute to key79

